src/codegen/subexpression_elimination/mod.rs

```diff
@@ -161,21 +209,24 @@ fn kill_loop_variables(block: &BasicBlock, cur_set: &mut AvailableExpressionSet,
         return;
     }
     for var_no in &block.loop_reaching_variables {
+        cur_set.remove_mapped(*var_no);
         cur_set.kill(*var_no);
     }
 }
 
 /// Find the correct visiting order for the CFG traversal, using topological sorting. The visiting
 /// order should be the same as the execution order. This function also returns a DAG for the
 /// execution graph. This helps us find the lowest common ancestor later.
-fn find_visiting_order(cfg: &ControlFlowGraph) -> (Vec<(usize, bool)>, Vec<Vec<usize>>) {
+fn find_visiting_order(cfg: &ControlFlowGraph) -> CfgAsDag {
```

Comment:
I did think the DAG could be created using BFS and then you would not need the stack, however if I remember correctly BFS requires a stack for its implementation so you it probably doesn't win you anything. :laughing:

Reply:
> The removal of cycles depends on the back edge being in the stack, so I guess that would have to be a dominating node. That is always true for code we generate, but surely it's possible to create code with gotos which make this fail (if we had gotos). Could do with a comment.

The algorithm works with gotos as well. A cycle happens when a vertex is in the stack but it hasn't been visited, when this happens the returning edge is removed. We can have gotos that do not create cycles (which would work as well) and those that create. If a go-to create a cycle, the detection would remove the returning edge too.

Reply:
> Also I am unsure why `degrees: Vec<i32>` uses a signed type. Can it go negative? If so, that should be explained.

While analyzing the graph, all degrees will be positive, because we I am not calculating out-degrees, only in-degrees. When calculating the topological sorting, I subtract 1 from the degree to determine which node I should be visiting next. This subtraction can make the value on the vector negative (which is the correct behavior).

https://github.com/hyperledger/solang/blob/12177ba59bca0e05113019e6a085d05d85ad9d32/src/codegen/subexpression_elimination/mod.rs#L196

Reply:
> The algorithm works with gotos as well. A cycle happens when a vertex is in the stack but it hasn't been visited. When this happens the returning edge is removed. We can have gotos that do not create cycles (which would work as well) and those that create. If a go-to creates a cycle, the detection would remove the returning edge too.

Yes, you are right. Nevermind